cartographer/mapping/map_builder.cc

```diff
@@ -177,8 +177,11 @@ void MapBuilder::LoadMap(io::ProtoStreamReader* const reader) {
   proto::SparsePoseGraph pose_graph;
   CHECK(reader->ReadProto(&pose_graph));
 
-  const int map_trajectory_id = AddTrajectoryForDeserialization();
-  sparse_pose_graph_->FreezeTrajectory(map_trajectory_id);
+  for (int trajectory_id = 0; trajectory_id < pose_graph.trajectory().size();
+       ++trajectory_id) {
+    CHECK_EQ(AddTrajectoryForDeserialization(), trajectory_id);
```

Comment:
It's a good idea to add trajectories with IDs 0, ..., n-1. The deserialized nodes and submaps have to be mapped to these trajectories then, though. Persisted trajectory IDs are `pose_graph.trajectory(trajectory_id).trajectory_id()`. One more suggestion: Make this a range-based for loop, remove this `CHECK`, and add a map from the persisted trajectory IDs to the new trajectory IDs. This way we would already support loading multiple maps.

Reply:
Oh, I did not have a remapping in mind, I just assumed that serialized trajectories would always be labeled from 0 to n-1. I am not a huge fan of remapping, because that would mean that we would have to also apply it when deserializing constraints. I would rather restore everything how it was previously.

Reply:
Sure, once all data from a trajectory is gone, the trajectory ID might not be there anymore. Moreover, the more important aspect is that we want to be able to deserialize multiple sources into one instance. This means there is no way around remapping during deserialization anyway.

Reply:
Makes sense. I will add the remapping.

Reply:
By the way: what do you think about applying the remapping to the SPG proto in-place, immediately? Or would you prefer to pass the remapping it to deserialization functions in the SPG (to be added in later PRs) which use the SPG proto?

Reply:
Not sure if this helps you a lot. You'd still have to create a map to efficiently look up the new trajectory IDs and keep it around since the nodes/submaps are in separate protos.

Reply:
Yes, but at least I wouldn't have to pass the map around - if I apply it the remapping to all protos in `LoadMap` (soon to become `LoadState`), I can hide the fact that a remapping was applied from the rest of Cartographer.

Reply:
I misunderstood and think we agree: if you pass a proto it makes sense to change it accordingly so that the consumers of that proto are not forced to implement remapping.